titus-server-master/src/main/java/com/netflix/titus/master/scheduler/constraint/TaskCache.java

```diff
@@ -53,36 +58,80 @@ public void prepare() {
-            this.zoneBalanceCountersByJobId = buildZoneBalanceCountersByJobId(jobsAndTasks);
+            this.assignedIpAllocations = new ConcurrentHashMap<>();
+            this.zoneBalanceCountersByJobId = new HashMap<>();
+            this.ipAllocationIdToZoneId = new HashMap<>();
+            buildTaskCacheInfo(jobsAndTasks);
         }
 
         private Map<String, Integer> getTasksByZoneIdCounters(String jobId) {
             return zoneBalanceCountersByJobId.getOrDefault(jobId, Collections.emptyMap());
         }
 
-        private Map<String, Map<String, Integer>> buildZoneBalanceCountersByJobId(List<Pair<Job, List<Task>>> jobsAndTasks) {
-            Map<String, Map<String, Integer>> result = new HashMap<>();
+        private void buildTaskCacheInfo(List<Pair<Job, List<Task>>> jobsAndTasks) {
             for (Pair<Job, List<Task>> jobAndTask : jobsAndTasks) {
                 Map<String, Integer> jobZoneBalancing = new HashMap<>();
                 for (Task task : jobAndTask.getRight()) {
                     String zoneId = getZoneId(task);
                     if (zoneId != null) {
                         jobZoneBalancing.put(zoneId, jobZoneBalancing.getOrDefault(zoneId, 0) + 1);
                     }
+
+                    getIpAllocationId(task).map(ipAllocationId -> {
+                        ipAllocationIdToZoneId.put(
+                                ipAllocationId,
+                                getIpAllocationZone(ipAllocationId, jobAndTask.getLeft().getJobDescriptor()).orElse(""));
+                        if (TaskState.isRunning(task.getStatus().getState())) {
+                            assignedIpAllocations.put(ipAllocationId, task.getId());
+                        }
+                        return null;
+                    });
                 }
-                result.put(jobAndTask.getLeft().getId(), jobZoneBalancing);
+                zoneBalanceCountersByJobId.put(jobAndTask.getLeft().getId(), jobZoneBalancing);
             }
-            return result;
         }
 
         private String getZoneId(Task task) {
             return task.getTaskContext().get(TaskAttributes.TASK_ATTRIBUTES_AGENT_ZONE);
         }
+
+        private Optional<String> getIpAllocationId(Task task) {
+            return Optional.ofNullable(task.getTaskContext().get(TaskAttributes.TASK_ATTRIBUTES_IP_ALLOCATION_ID));
+        }
+
+        private Optional<String> getIpAllocationZone(String ipAllocationId, JobDescriptor<?> jobDescriptor) {
+            for (SignedIpAddressAllocation signedIpAddressAllocation : jobDescriptor.getContainer().getContainerResources().getSignedIpAddressAllocations()) {
+                if (signedIpAddressAllocation.getIpAddressAllocation().getAllocationId().equals(ipAllocationId)) {
+                    return Optional.of(signedIpAddressAllocation.getIpAddressAllocation().getIpAddressLocation().getAvailabilityZone());
+                }
+            }
+            return Optional.empty();
```

Comment:
Or there was a bug in how the IP Allocation ID was assigned to the Task. This would cause `IpAllocationConstraint` to return `IP_ALLOCATION_NOT_IN_ZONE` because the map would contain `""` as the zone and fail the match the instance's zone. But, yes, same effect of not being schedulable.

I think that's a good idea. The next PR is to add a tastContext for "waiting for an in use IP" case and I can handle this in the same way when we evaluate the scheduling failure and add the task context you mentioned. Let me know if you're not OK with it being in the next PR.

Reply:
Next PR is fine